spopt/tests/test_p_dispersion.py

```diff
@@ -0,0 +1,162 @@
+
+        self.facilities_snapped = spaghetti.element_as_gdf(
+            ntw, pp_name="facilities", snapped=True
+        )
+
+        self.cost_matrix = ntw.allneighbordistances(
+            sourcepattern=ntw.pointpatterns["facilities"],
+            destpattern=ntw.pointpatterns["facilities"],
+        )
+
+    def test_p_dispersion_from_cost_matrix(self):
+        pdispersion = PDispersion.from_cost_matrix(self.cost_matrix, p_fac=2)
+        result = pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+        self.assertIsInstance(result, PDispersion)
+
+    def test_p_dispersion_from_geodataframe(self):
+        pdispersion = PDispersion.from_geodataframe(
+            self.facilities_snapped,
+            "geometry",
+            p_fac=2,
+        )
+        result = pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+        self.assertIsInstance(result, PDispersion)
+
+
+class TestRealWorldLocate(unittest.TestCase):
+    def setUp(self) -> None:
+        self.dirpath = os.path.join(os.path.dirname(__file__), "./data/")
+        network_distance = pandas.read_csv(
+            self.dirpath
+            + "SF_network_distance_candidateStore_16_censusTract_205_new.csv"
+        )
+
+        ntw_dist_piv = network_distance.pivot_table(
+            values="distance", index="DestinationName", columns="name"
+        )
+
+        self.cost_matrix = ntw_dist_piv.to_numpy()
+
+        facility_points = pandas.read_csv(self.dirpath + "SF_store_site_16_longlat.csv")
+
+        self.facility_points_gdf = (
+            geopandas.GeoDataFrame(
+                facility_points,
+                geometry=geopandas.points_from_xy(
+                    facility_points.long, facility_points.lat
+                ),
+            )
+            .sort_values(by=["NAME"])
+            .reset_index()
+        )
+
+        self.service_dist = 5000.0
+        self.p_facility = 4
+
+    def test_optimality_p_dispersion_from_cost_matrix(self):
+        pdispersion = PDispersion.from_cost_matrix(
+            self.cost_matrix, p_fac=self.p_facility
+        )
+        pdispersion = pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+        self.assertEqual(pdispersion.problem.status, pulp.LpStatusOptimal)
+
+    # how to force this to fail???
+    def test_infeasibility_p_dispersion_from_cost_matrix(self):
+        pdispersion = PDispersion.from_cost_matrix(self.cost_matrix, p_fac=0)
+        with self.assertRaises(RuntimeError):
+            pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+
+    def test_optimality_p_dispersion_from_geodataframe(self):
+        pdispersion = PDispersion.from_geodataframe(
+            self.facility_points_gdf,
+            "geometry",
+            p_fac=self.p_facility,
+        )
+        pdispersion = pdispersion.solve(pulp.PULP_CBC_CMD(msg=False))
+        self.assertEqual(pdispersion.problem.status, pulp.LpStatusOptimal)
+
+    # how to force this to fail??? a cost matrix with all facilities equidistant?
```

Comment:
I think this would indicate a *degenerate* problem (one with no right answer) rather than one where the solution would be infeasible. I'm not sure that there is an infeasibility condition for p-dispersion, unless you ask for *p* sites and you only have less than *p* candidates.

Reply:
@ljwolf This is a good point that I overlooked.

Reply:
You're correct! Assigning a p value that exceeds the number of candidate facilities seems to be the best approach. Thanks!